### PR TITLE
R4R: Add gateway edition

### PR DESCRIPTION
**Pull request by secret2830**
#1497

### CODE REVIEW

File: app/v1/asset/keeper.go [hunk 1]
```diff
@@ -76,8 +78,39 @@ func (k Keeper) CreateGateway(ctx sdk.Context, msg MsgCreateGateway) (sdk.Tags,
...
+	}
+
+	// set the new gateway
+	k.SetGateway(ctx, gateway)
+
 	// TODO
-	return nil, nil
+	editTags := sdk.NewTags(
+		"id", []byte{gateway.ID},
+		"moniker", []byte(msg.Moniker),
```
Comment (chengwenxi) [inline app/v1/asset/keeper.go]: Consider defining tag's key in tags.go?

Comment (secret2830) [reply app/v1/asset/keeper.go]: Tags handling is put inot TODO lists. Entire asset tags are to be designed later

File: app/v1/asset/keeper.go [hunk 2]
```diff
@@ -76,8 +78,39 @@ func (k Keeper) CreateGateway(ctx sdk.Context, msg MsgCreateGateway) (sdk.Tags,
...
+	if !msg.Owner.Equals(gateway.Owner) {
+		return nil, ErrInvalidOwner(k.codespace, fmt.Sprintf("the address %d is not the owner of the gateway %s", msg.Owner, msg.Moniker))
+	}
+
+	// update the gateway
+	gateway.Identity = msg.Identity
+	gateway.Details = msg.Details
+	gateway.Website = msg.Website
+	gateway.RedeemAddr = msg.RedeemAddr
+	gateway.Operators = msg.Operators
```
Comment (chengwenxi) [inline app/v1/asset/keeper.go]: It is difficult to use if you only want to change one parameter.

File: app/v1/asset/keeper.go [hunk 3]
```diff
@@ -62,8 +62,10 @@ func (k Keeper) CreateGateway(ctx sdk.Context, msg MsgCreateGateway) (sdk.Tags,
 		Operators:  msg.Operators,
 	}
 
-	// save the gateway with the creation enabled
-	k.saveGateway(ctx, gateway, true)
+	// set the gateway and related keys
+	k.SetGateway(ctx, gateway)
+	k.SetMoniker(ctx, msg.Moniker, gatewayID)
+	k.SetOwnerGatewayID(ctx, msg.Owner, gatewayID)
```
Comment (zhangyelong) [inline app/v1/asset/keeper.go]: Does this mean that one owner can only create one gateway ?

File: app/v1/asset/keeper.go (hunk 2 above)
Comment (secret2830) [reply app/v1/asset/keeper.go]: > It is difficult to use if you only want to change one parameter.

it's up to clients to fill the existing values.

File: app/v1/asset/keeper.go (hunk 3 above)
Comment (secret2830) [reply app/v1/asset/keeper.go]: > Does this mean that one owner can only create one gateway ?

No. This key is used to iterate ids on an owner with prefix. Its format is "gateways:<owner>:<id>"

File: app/v1/asset/keeper.go (hunk 2 above)
Comment (haifengxi) [inline app/v1/asset/keeper.go]: Should allow user to specify values for only those fields they want to update.  Refer to `stake/keeper/validator.go` or come up with something even better.

Comment (secret2830) [reply app/v1/asset/keeper.go]: Done

**Pull request merged**